### PR TITLE
Implement MarkDistinctOperator

### PR DESCRIPTION
**Pull request by shiyu-bytedance**
The MarkDistinctOperator accepts a vector as input outputs the input vector with an additional "MarkDistinct" boolean column as output.

For the first time each unique combination of select channels from the input vector is detected, the "MarkDistinct" output column records true, otherwise, false.

For example,

Input:
c0, c1, c2
0, 1, 2
1, 1, 2
2,1, 1
0, 1, 2
1, 3, 0
2,1,1

Select channels: c1, c2

Output:
c0, c1, c2, isDistinct
0, 1, 2, true
1, 1, 2, false
2,1, 1, true
0, 1, 2, false
1, 3, 0, true
2,1,1, false

This is needed for TCP-DS

### CODE REVIEW

**Comment by shiyu-bytedance** [comment]
@mbasmanova hello again, I implemented MarkDistinctNode and would like to contribute to velox.

**Comment by kagamiori** [comment]
Hi @shiyu-bytedance, Glad to have your contribution! Could you add a description to this PR about what this operator does and why it is needed? Thanks!

**Comment by shiyu-bytedance** [comment]
> 

Done. https://github.com/facebookincubator/velox/pull/2321#issue-1342651362

File: velox/exec/tests/utils/PlanBuilder.h [hunk 1]
```diff
@@ -520,6 +524,7 @@ class PlanBuilder {
   /// the input. Output columns may appear in different order from the input,
   /// some input columns may be missing in the output, some columns may be
   /// duplicated in the output.
+
```
Comment (aditi-pandit) [inline velox/exec/tests/utils/PlanBuilder.h:527]: Please remove newline.

File: velox/exec/tests/utils/PlanBuilder.h [hunk 2]
```diff
@@ -505,6 +505,10 @@ class PlanBuilder {
       const std::string& idName = "unique",
       const int32_t taskUniqueId = 1);
 
+  PlanBuilder& markDistinct(
+      const std::vector<std::shared_ptr<const core::FieldAccessTypedExpr>>&
```
Comment (aditi-pandit) [inline velox/exec/tests/utils/PlanBuilder.h:509]: Use alias "FieldAccessTypedExprPtr"

File: velox/exec/tests/utils/PlanBuilder.h [hunk 3]
```diff
@@ -505,6 +505,10 @@ class PlanBuilder {
...
+          distinctVariables);
```
Comment (aditi-pandit) [inline velox/exec/tests/utils/PlanBuilder.h:510]: Rename "distinctFIelds" ?

File: velox/core/PlanNode.h [hunk 4]
```diff
@@ -1565,4 +1565,51 @@ class WindowNode : public PlanNode {
...
+/// Uses:
+///
+/// to create a masking flag for use with a DISTINCT aggregator.
+/// in scalar correlated subqueries
+
+class MarkDistinctNode : public PlanNode {
+ public:
+  MarkDistinctNode(
+      const PlanNodeId& id,
+      const std::vector<std::shared_ptr<const FieldAccessTypedExpr>>&
```
Comment (aditi-pandit) [inline velox/core/PlanNode.h:1580]: Use alias "FieldAccessTypedExprPtr"

File: velox/core/PlanNode.h [hunk 5]
```diff
@@ -1565,4 +1565,51 @@ class WindowNode : public PlanNode {
...
+          distinctVariables,
```
Comment (aditi-pandit) [inline velox/core/PlanNode.h:1581]: Nit : Rename distinctFields. Change this naming uniformly across this class.

File: velox/core/PlanNode.h [hunk 6]
```diff
@@ -1565,4 +1565,51 @@ class WindowNode : public PlanNode {
...
+  const RowTypePtr& inputType() const {
+    return sources_[0]->outputType();
+  }
+
+  std::string_view name() const override {
+    return "MarkDistinct";
+  }
+
+  const std::vector<std::shared_ptr<const core::ITypedExpr>>
+  getDistinctVariables() const {
```
Comment (aditi-pandit) [inline velox/core/PlanNode.h:1601]: Why did you use ITYpedExpr here instead of FIeldAccessTypedExpr ?

File: velox/exec/LocalPlanner.cpp [hunk 7]
```diff
@@ -411,6 +412,14 @@ std::shared_ptr<Driver> DriverFactory::createDriver(
           assignUniqueIdNode,
           assignUniqueIdNode->taskUniqueId(),
           assignUniqueIdNode->uniqueIdCounter()));
+    } else if (
+        auto markDistinctNode =
+            std::dynamic_pointer_cast<const core::MarkDistinctNode>(planNode)) {
+      std::vector<uint32_t> channels = toChannels(
```
Comment (aditi-pandit) [inline velox/exec/LocalPlanner.cpp:418]: We can do this logic inside the constructor of the MarkDistinct operator. The distinct fields can be directly obtained from the PlanNode there, so there isn't a need for the extra variable for the input channels.

File: velox/exec/CMakeLists.txt [hunk 8]
```diff
@@ -55,6 +55,8 @@ add_library(
   Unnest.cpp
   Values.cpp
   VectorHasher.cpp
+  AssignUniqueId.cpp
```
Comment (aditi-pandit) [inline velox/exec/CMakeLists.txt:58]: This file is already present as the last entry in the list. Please remove it.

File: velox/exec/MarkDistinct.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  // Decode the channels.
+  std::fill(columnExists_.begin(), columnExists_.end(), false);
+  for (auto idx = 0; idx < keyChannels_.size(); ++idx) {
+    if (input->childAt(keyChannels_[idx])) {
+      columnExists_[idx] = true;
+      decodedVectors_[idx].decode(*input->children()[idx], allRows_);
+    }
+  }
+
+  VectorPtr& result = results_[0];
```
Comment (aditi-pandit) [inline velox/exec/MarkDistinct.cpp:61]: It seems this logic is looking for distinct rows within this input buffer itself. I thought we would need to look globally. Are you assuming that the input is sorted so a local check is sufficient ?

File: velox/exec/MarkDistinct.h [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  RowVectorPtr getOutput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /*future*/) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override;
+
+ private:
+  std::unordered_set<uint64_t> hashSet_;
```
Comment (aditi-pandit) [inline velox/exec/MarkDistinct.h:54]: Please can you add comments for the meaning of these fields.

File: velox/exec/tests/CMakeLists.txt [hunk 11]
```diff
@@ -55,7 +55,9 @@ add_executable(
   TreeOfLosersTest.cpp
   UnnestTest.cpp
   VectorHasherTest.cpp
-  WindowFunctionRegistryTest.cpp)
+  WindowFunctionRegistryTest.cpp
+  AssignUniqueIdTest.cpp
```
Comment (aditi-pandit) [inline velox/exec/tests/CMakeLists.txt:59]: This test is present towards the beginning of this list. Please remove this entry.

File: velox/exec/tests/MarkDistinctTest.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::exec::test;
+
+class MarkDistinctTest : public OperatorTestBase {
+ protected:
+  template <class NativeType>
+  std::function<VectorPtr()> baseFunctorTemplate() {
+    return [&]() {
+      return makeFlatVector<NativeType>(
```
Comment (aditi-pandit) [inline velox/exec/tests/MarkDistinctTest.cpp:31]: Please can you add more tests that cover multiple batches of input rows.

**Review by aditi-pandit — COMMENTED** [review]
Thanks @shiyu-bytedance for this code. Took a quick look at your code. Overall looks okay. Had some questions.

File: velox/exec/MarkDistinct.cpp (hunk 9 above)
Comment (aditi-pandit) [reply velox/exec/MarkDistinct.cpp:61]: Nvm... Noticed that your hashSet_ structure is global so should take care of this. Please add relevant comments about all these member variables in the .h file for this.

File: velox/core/PlanNode.cpp [hunk 13]
```diff
@@ -419,6 +419,24 @@ void AssignUniqueIdNode::addDetails(std::stringstream& /* stream */) const {
...
+  distinctVariables_ = std::move(distinctVariables);
+  std::vector<std::string> names(sources_[0]->outputType()->names());
+  std::vector<TypePtr> types(sources_[0]->outputType()->children());
+  names.emplace_back("isDistinct");
+  types.emplace_back(BOOLEAN());
+  outputType_ = ROW(std::move(names), std::move(types));
+}
+
+void MarkDistinctNode::addDetails(std::stringstream& /* stream */) const {
+  // Nothing to add.
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:437]: Shouldn't we specify the "distinctVariables" here?

File: velox/exec/MarkDistinct.h (hunk 10 above)
Comment (mbasmanova) [inline velox/exec/MarkDistinct.h:54]: Any particular reason not to use VectorHasher and a HashTable? hashSet is slow.

File: velox/exec/MarkDistinct.cpp [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+          auto& decodedChild = decodedVectors_[idx];
+          auto cellHash =
+              decodedChild.base()->hashValueAt(decodedChild.indices()[row]);
+          hash = isFirst ? cellHash : bits::hashMix(hash, cellHash);
+          isFirst = false;
+        }
+      }
+    }
+
+    if (hashSet_.find(hash) == hashSet_.end()) {
```
Comment (mbasmanova) [inline velox/exec/MarkDistinct.cpp:82]: What about hash collisions?

File: velox/exec/tests/utils/PlanBuilder.h (hunk 2 above)
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.h:509]: Please, document this method and change input to std::vector<std::string> to allow for passing column names like in other methods.

**Review by mbasmanova — CHANGES_REQUESTED** [review]
@shiyu-bytedance Shiyu, thank you for adding a new operator. Please, add it to documentation: velox/docs/develop/operators.rst

Aditi left some good comments. I added a few more.

**Pull request closed without merging**